File: tslox/interpreter/expr.ts
import {Token} from "./token";

export type Expr = Binary | Grouping | Literal | Unary

export class Binary {
    constructor(public left: Expr, public operator: Token, public right: Expr){}

    public accept<T>(visitor: Visitor<T>): T{
        return visitor.visitBinaryExpr(this);
    }
}

export class Grouping {
    constructor(public expression: Expr){}

    public accept<T>(visitor: Visitor<T>): T{
        return visitor.visitGroupingExpr(this);
    }
}

export class Literal {
    constructor(public value: Object){}

    public accept<T>(visitor: Visitor<T>): T{
        return visitor.visitLiteralExpr(this);
    }
}

export class Unary {
    constructor(public operator: Token, public right: Expr){}

    public accept<T>(visitor: Visitor<T>): T{
        return visitor.visitUnaryExpr(this);
    }
}

export interface Visitor<T> {
    visitBinaryExpr(expr: Binary): T
    visitGroupingExpr(expr: Grouping): T
    visitLiteralExpr(expr: Literal): T
    visitUnaryExpr(expr: Unary): T
}

File: pylox/pylox/expr.py
class Binary:
    def __init__(self, left, operator, right):
        self.left = left
        self.operator = operator
        self.right = right

    def accept(self, visitor):
        return visitor.visit_binary_expr(self)


class Grouping:
    def __init__(self, expression):
        self.expression = expression

    def accept(self, visitor):
        return visitor.visit_grouping_expr(self)


class Literal:
    def __init__(self, value):
        self.value = value

    def accept(self, visitor):
        return visitor.visit_literal_expr(self)


class Unary:
    def __init__(self, operator, right):
        self.operator = operator
        self.right = right

    def accept(self, visitor):
        return visitor.visit_unary_expr(self)
